### PR TITLE
gergo/web 2109 project region based db connection selector

### PR DESCRIPTION
**Pull request by gjedlicska**
- feat(projects): add project regions, default to null
- feat(multiregion): add projectRegion Db client lookup logic
- feat(multiregion): add project region repositories and caching
- feat(multiRegion): db initialization and get project db client
- feat(docker-compose): add second db for regions testing
- feat(multiRegion): initialize region with pubs and subs working
- fix(multiRegion): get region client even if it was registered in another pod
- feat(workspaces): create workspace resolver split
- feat(projects): rewrite project creation
- feat(multiRegion): getRegionDb

## Checklist:

- [ ] My pull request follows the guidelines in the [Contributing guide](https://github.com/specklesystems/speckle-server/blob/main/CONTRIBUTING.md)?
- [ ] My pull request does not duplicate any other open [Pull Requests](../../pulls) for the same update/change?
- [ ] My commits are related to the pull request and do not amend unrelated code or documentation.
- [ ] My code follows a similar style to existing code.
- [ ] I have added appropriate tests.
- [ ] I have updated or added relevant documentation.

### CODE REVIEW

File: packages/server/assets/multiregion/typedefs/multiregion.graphql [hunk 1]
```diff
@@ -2,7 +2,7 @@ type ServerRegionItem {
   id: String!
   key: String!
   name: String!
-  description: String!
+  description: String
```
Comment (fabis94) [inline packages/server/assets/multiregion/typedefs/multiregion.graphql:5]: Do we throw if creating a workspace invite using old mutation?

File: packages/server/modules/core/domain/projects/logic.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+const adjectives = [
```
Comment (fabis94) [inline packages/server/modules/core/domain/projects/logic.ts:1]: Isn't this just a helper? Why put it in the `domain` folder? + The name is too generic (logic.ts)

File: packages/server/modules/core/services/projects.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      regionKey: regionKey || null,
+      allowPublicComments: false
+    }
+
+    await storeProject({ project })
+    const projectId = project.id
+    // if regionKey, we need to make sure it is actually written and synced
+    if (regionKey) {
+      try {
+        await retry(
```
Comment (fabis94) [inline packages/server/modules/core/services/projects.ts:61]: we've already got retry in @speckle/shared fyi

**Review by fabis94 — CHANGES_REQUESTED** [review]
Left comments on some relatively minor issues. Am I assuming correctly that after this is merged the only (current) way to run any any of this multiregion logic is if you have those custom DO DBs prepared for you?

File: packages/server/modules/workspaces/graph/resolvers/workspaces.ts [hunk 4]
```diff
@@ -740,6 +756,59 @@ export = FF_WORKSPACES_MODULE_ENABLED
...
+              await createBranchFactory({ db })({
+                authorId,
+                description,
+                name,
+                streamId: projectId
+              })
+            }
+
+          // todo, use the command factory here, but for that, we need to migrate to the event bus
+          const createNewProject = createNewProjectFactory({
```
Comment (fabis94) [inline packages/server/modules/workspaces/graph/resolvers/workspaces.ts:794]: Can't say I'm a fan of the DB selection logic being left up to the dev writing the DI code. You can accidentally use the wrong DB here and fuck shit up for real. I guess if we get a DI container in place its going to be better.

I guess you could also build a factory w/ just 2 deps - db and project db, and then that factory could correctly build createNewProject and all related factories.

File: setup/db/9-docker_postgres_aiven.sql [hunk 5]
```diff
@@ -0,0 +1,2 @@
+ALTER SYSTEM SET wal_level = logical;
```
[the remark this answers is not in the document]
Comment (gjedlicska) [reply setup/db/9-docker_postgres_aiven.sql:1]: no

File: packages/server/modules/workspaces/graph/resolvers/workspaces.ts (hunk 4 above)
Comment (gjedlicska) [reply packages/server/modules/workspaces/graph/resolvers/workspaces.ts:794]: guys, lets not optimize for things before we know the pattern.

This is the first go at a multi region service instantiation. I do not know, what will work, but from the prev trials with the DI container, this is very possible to be centralized and codified in one place.

Until  then, I'm going to type these out to see what works and what doesn't

File: packages/server/assets/multiregion/typedefs/multiregion.graphql (hunk 1 above)
Comment (gjedlicska) [reply packages/server/assets/multiregion/typedefs/multiregion.graphql:5]: 1. we should, but i have not yet done that
2. its not yet updated, that is why i asked for help in the morning,

File: packages/server/modules/core/services/projects.ts (hunk 3 above)
Comment (gjedlicska) [reply packages/server/modules/core/services/projects.ts:61]: i know, but it does not have exponential backoff

File: docker-compose-deps.yml [hunk 6]
```diff
@@ -13,11 +13,29 @@ services:
       POSTGRES_PASSWORD: speckle
     volumes:
       - postgres-data:/var/lib/postgresql/data/
+      - ./setup/db/0-docker_postgres_aiven.sql:/docker-entrypoint-initdb.d/9-docker_postgres_aiven.sql
       - ./setup/db/10-docker_postgres_init.sql:/docker-entrypoint-initdb.d/10-docker_postgres_init.sql
       - ./setup/db/11-docker_postgres_keycloack_init.sql:/docker-entrypoint-initdb.d/11-docker_postgres_keycloack_init.sql
     ports:
       - '127.0.0.1:5432:5432'
 
+  postgres-region1:
```
[the remark this answers is not in the document]
Comment (gjedlicska) [reply docker-compose-deps.yml:22]: as i said in the morning, i could not get this working, due to the docker private network sync issue

[the remark this answers is not in the document]
Comment (gjedlicska) [reply docker-compose-deps.yml:22]: added one

**Pull request merged**